keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1182 @@
+- keep binary-version 1.31 while upgrading compat-version to 1.31 (stepwise upgrade of compatibility)
+
+Benefits to upgrading binary version independent of compatibility version:
+
+- A skip-version binary upgrade that transitions through compat versions has
+  higher probability of bugs in those intermediate binary versions already being
+  fixed.
+- During an upgrade, it is possible upgrade the new binary version while the
+  compatibility version is fixed (e.g. `binaryVersion: 1.28` -> `{binaryVersion: 1.29, compatVersion: 1.28}`).
+  This allow differences in the binary (bugs fixed or introduced, performance
+  changes, ...) to be introduced into a cluster and verified safe before
+  allowing access to the new APIs and features new version. Once the binary
+  version has been deamed safe, the compatibility version can then be upgraded.
+- Any upgrade system that successfully detects failures between upgrade steps
+  can report which upgrade step failed. This makes it easier to diagnose the
+  failures, because there are fewer possible causes of the failure. (There's a
+  huge difference between "A cluster upgrade failed" and "A cluster upgrade
+  failed when only the apiserver's binary version changed").
+- For upgrades where multiple failures occur, this increases the odds those
+  failures are split across steps. An upgrade system that is able to pause after
+  a step where failures are detected allow for failures at that step to be
+  addressed before proceeding to subsequent steps. These failures can be
+  addressed without the disruption and "noise" from failures in subsequent
+  steps.
+- A compatibility version rollback can be performed without changing binary
+  version.
+
+
+### Goals
+
+- Introduce the metadata necessary to configure features/APIs/storage-versions/CEL
+  features to match the behavior of an older Kubernetes release version
+- A Kubernetes binary with compatibility version set to N, will pass the
+  conformance and e2e tests from Kubernetes release version N.
+- A Kubernetes binary with compatibility version set to N does not enable any
+  changes (storage versions, CEL feature, features) that would prevent it
+  from being rolled back to N-1.
+- The most recent Kubernetes version supports compatibility version being set to
+  the full range of supported versions (N..N-3).
+
+### Non-Goals
+
+- Support compatibility version for Alpha features.  Alpha feature are not
+  designed to be upgradable, so we will not allow alpha features to be used in
+  conjunction with compatibility version.
+- Changes to Cluster API/kubeadm/KIND/minikube to absorb the compatibility versions
+  will be addressed separate from this KEP
+
+## Proposal
+
+### Component Flags
+
+Kubernetes components (apiservers, controller managers, schedulers) will offer a
+`--compatibility-version` flag that can be set to any of the previous three
+minor versions. If unset, the compatibility version defaults to the `<major.minor>`
```

Comment:
> `--compatibility-version` flag that can be set to any of the previous three minor versions.

I was expecting a first phase where the only options were `1.n` and `1.(n-1)`

That would let us get the version tracking and mechanics in place as a first step, without having to support additional skew immediately. I'm not actually confident all the gates required to support back three versions still exist, so putting the mechanics in place and then letting new releases age us into supporting 3 versions back would be easier to accomplish.

Reply:
Yes, the initial offering would be `1.n` and `1.(n-1)`. After we establish the best practice for compatibility versions, the long term goal is to expand it to the previous three minor versions.

Reply:
I've swept the KEP and clarified this.